商品購入機能

sayaka-ando opened the pull request:
# レビュー依頼
# 商品購入機能実装が完了したため

- 必要な情報を適切に入力すると、商品の購入ができること
- 購入が完了したら、トップページまたは購入完了ページに遷移すること
- 商品購入ページでは、一覧や詳細ページで選択した商品の情報が出力されること
- クレジットカード決済ができること
https://gyazo.com/96ce9e5af85c4d18e492b32c4c941777

- ログアウト状態のユーザーは、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移すること

- ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移すること
https://gyazo.com/2cebbf42a041e679fe74c2728034f0ca

- ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移すること
https://gyazo.com/5bce9ea631e42b29ab675e068bc34363

- クレジットカードの情報は購入の都度入力させること
- クレジットカード情報は必須であり、正しいクレジットカードの情報で無いときは決済できないこと
- 配送先の住所情報も購入の都度入力させること
- 入力に問題がある状態で購入ボタンが押されたら、購入ページに戻りエラーメッセージが表示されること
https://gyazo.com/6f574cec94cb9ba182992ef479a1816a

- 配送先の情報として、郵便番号・都道府県・市区町村・番地・電話番号が必須であること
- 郵便番号にはハイフンが必要であること（123-4567となる）
- 電話番号にはハイフンは不要で、11桁以内であること（09012345678となる）
https://gyazo.com/7bd1575c6e9e15f068991958c76f9835

sayaka-ando committed:
formモデルの作成

sayaka-ando committed:
モデルの作成とformオブジェクトの作成

sayaka-ando committed:
コントローラーの記述

sayaka-ando committed:
formオブジェクトのテストコード

sayaka-ando committed:
トークンの作成、送信（js）

sayaka-ando committed:
テストコードと環境変数の設定

sayaka-ando committed:
購入ページへの条件分岐

chisesakai commented:
レビュー依頼ありがとうございます。
Gyazoのキャプチャのご提出ありがとうございます。項目ごとに整理されていて非常に確認しやすかったです！

#### 確認させていただきたい挙動があるため、下記3点のキャプチャも追加で頂けますでしょうか？
- ログアウト状態のユーザーは、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移すること(項目としては頂いておりますが、添付のURLを記述頂いていないようです。)
- 出品者・出品者以外にかかわらず、ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画(編集機能要件より)
- ログイン状態の出品者でも、売却済みの商品に対しては「編集・削除ボタン」が表示されない動画（詳細機能要件より)

恐れ入りますが、上記添付の上、再度レビュー依頼のご提出をお願い致します。

sayaka-ando committed:
ログイン状況、出品状況での条件分岐

• ログアウト状態のユーザーは、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移すること

• 出品者・出品者以外にかかわらず、ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画(編集機能要件より)

• ログイン状態の出品者でも、売却済みの商品に対しては「編集・削除ボタン」が表示されない動画（詳細機能要件より)

sayaka-ando commented:
# 確認お願いします
# 修正が完了したため

確認、ご指摘ありがとうございます！
下記、お願いします。sold out表示も実装完了しました。

• ログアウト状態のユーザーは、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移すること
https://gyazo.com/1f8c0d1f1a547e56087c8e619d6a5f43

• 出品者・出品者以外にかかわらず、ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画(編集機能要件より)
https://gyazo.com/b0f54d6b1b8e1417dfd4896479bce829

• ログイン状態の出品者でも、売却済みの商品に対しては「編集・削除ボタン」が表示されない動画（詳細機能要件より)
https://gyazo.com/5f5fa87ee4c62de3a961009066b0aa9e

suginaka commented on app/models/order_form.rb:
```diff
@@ -0,0 +1,18 @@
+  validates :address, presence: true
+  validates :phone_number, presence: true, length: { maximum: 11 }, format: { with: /\A[0-9]+\z/ }
+  validates :token, presence: true
+
```
user_id、item_idにもバリデーションpresence: trueを追加しましょう。

【理由】
 今まではモデル間のアソシエーションを組むことで、自動的にpresence: trueが働いておりました。
 ただ、フォームオブジェクトでは直接アソシエーションを組んでいないためにバリデーションが必要となります。

suginaka commented on app/models/order_form.rb:
```diff
@@ -0,0 +1,18 @@
+class OrderForm
+  include ActiveModel::Model
+
```
バリデーションは「with_options」を使用することで、共通したオプションを付けることが可能となりますので、共通のバリデーションをまとめましょう。

suginaka commented on spec/factories/order_forms.rb:
```diff
@@ -0,0 +1,10 @@
+    postal_code       { '000-0000' }
+    delivery_area_id  { 1 }
+    city              { '名古屋市' }
+    address           { '1丁目' }
```
建物のカラムを追加しましょう。

suginaka commented on spec/models/order_form_spec.rb:
```diff
@@ -0,0 +1,72 @@
+        expect(@order_form).to be_valid
+      end
+    end
+
```
「建物名は空でも登録できる」のテストを追加しましょう。

Hide-Morishita replied:
「建物名は空でも登録できる」のテストを追加しましょう。

suginaka commented on spec/models/order_form_spec.rb:
```diff
@@ -0,0 +1,72 @@
+        @order_form.valid?
+        expect(@order_form.errors.full_messages).to include("Token can't be blank")
+      end
+
```
```
user_idが空では登録できない
item_idが空では登録できない
```
のテストを追加しましょう。

Hide-Morishita replied:
```
user_idが空では登録できない
item_idが空では登録できない
```
のテストを追加しましょう。

suginaka reviewed (commented):
レビュー依頼ありがとうございます。
確認いたしました！
コメントしましたので、ご確認お願いします。

sayaka-ando committed:
テストコードの追加

sayaka-ando commented:
# 確認お願いします
# 修正が完了したため

指摘ありがとうございました。
修正しましたのでご確認お願いします。
https://gyazo.com/578a9a4cfbed4d97c87be0fb028ecff4

Hide-Morishita commented on app/controllers/order_informations_controller.rb:
```diff
@@ -0,0 +1,40 @@
+class OrderInformationsController < ApplicationController
+  before_action :authenticate_user!
+
+  def index
+    @item = Item.find(params[:item_id])
+    if current_user.id == @item.user_id || @item.order_information.present?
+      redirect_to root_path
+    end
+    @order_form = OrderForm.new
+    
+  end
+
+  def create
+    @item = Item.find(params[:item_id])
+    @order_form = OrderForm.new(order_information_params)
+      if @order_form.valid?
+        pay_item
+        @order_form.save
+        redirect_to root_path
+      else
+        render action: :index
+      end
+  end
```
```ruby
@item = Item.find(params[:item_id])
```
こちらの記述ですが、indexとcreateにて同じ記述がありますので、メソッドとして記述しbefore_actionで呼び出すようにしましょう。

Hide-Morishita commented on app/models/order_form.rb:
```diff
@@ -0,0 +1,18 @@
+  validates :city, presence: true
+  validates :address, presence: true
+  validates :phone_number, presence: true, length: { maximum: 11 }, format: { with: /\A[0-9]+\z/ }
+  validates :token, presence: true
```
presence: trueの記述が重複しています。
バリデーションは「with_options」を使用することで、共通したオプションを付けることが可能となりますので,
```ruby
with_options presence: true do
 validates :○○
end
```
のように修正いたしましょう！
またuser_id,item_idもバリデーションの記述をしましょう。

Hide-Morishita commented on spec/factories/addresses.rb:
```diff
@@ -0,0 +1,5 @@
+FactoryBot.define do
+  factory :address do
+    
+  end
+end
```
こちら不要なファイルであれば削除しておきましょう。

Hide-Morishita commented on spec/factories/order_informations.rb:
```diff
@@ -0,0 +1,3 @@
+FactoryBot.define do
+  
+end
```
こちら不要なファイルであれば削除しておきましょう。

Hide-Morishita commented on spec/models/order_information_spec.rb:
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe OrderInformation, type: :model do
+  
+end
```
こちら不要なファイルであれば削除しておきましょう。

Hide-Morishita commented on spec/requests/addresses_request_spec.rb:
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe "Addresses", type: :request do
+
+end
```
こちら不要なファイルであれば削除しておきましょう。

Hide-Morishita reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いいたします。

sayaka-ando committed:
before_actioの設定とファイルの削除

sayaka-ando commented:
# 確認お願いします
# 修正が完了したため
指摘ありがとうございました。
修正しましたのでご確認お願いします。
https://gyazo.com/578a9a4cfbed4d97c87be0fb028ecff4

takahirohoriguchi commented:
レビュー依頼ありがとうございます！
キャプチャーもご丁寧にありがとうございます！

LGTMとさせていただきます！本当にお疲れ様でした！
最後にマージとプルだけ忘れずにお願いします！
